bug: router crashes at around 4000 tokens prompt

### Describe the bug

Hello,

**Issue:**
I was testing the Semantic Router using the dashboard "Playground" and "Brain" sections and I found out that when I place a prompt around 4000 tokens the router crashes without any errors in the container.


**Suggestion:**
I dug deeper and found out that there are 4 places where it can crash without providing any logs (Keep in mind that this research was done with AI, I did not validate it):
1. pkg/classification/authz_classifier.go:157
2. pkg/extproc/processor_core.go:49 - no recover() wrapper
3.  Several goroutines lacking panic recovery:
   - pkg/extproc/server.go:437 — config reload goroutine
   - pkg/extproc/processor_res_memory.go:27 — memory store goroutine
   - pkg/extproc/req_filter_modality.go:270,277 — parallel AR text + diffusion calls
   - pkg/extproc/router_selection.go:375 — lookup table populator

### To Reproduce

`"2026-04-28T08:14:12.762Z [DEBUG] Watching for changes in setting files C:\Users\x\.claude...
2026-04-28T08:14:12.767Z [DEBUG] [init] configureGlobalMTLS starting
2026-04-28T08:14:12.768Z [DEBUG] [init] configureGlobalMTLS complete
2026-04-28T08:14:12.768Z [DEBUG] [init] configureGlobalAgents starting
2026-04-28T08:14:12.771Z [DEBUG] [init] configureGlobalAgents complete
2026-04-28T08:14:12.812Z [DEBUG] Using bash path: "C:\Program Files\Git\bin\bash.exe"
2026-04-28T08:14:12.816Z [DEBUG] [LSP MANAGER] initializeLspServerManager() called
2026-04-28T08:14:12.816Z [DEBUG] [LSP MANAGER] Created manager instance, state=pending
2026-04-28T08:14:12.817Z [DEBUG] [LSP MANAGER] Starting async initialization (generation 1)
2026-04-28T08:14:12.820Z [DEBUG] [STARTUP] Loading MCP configs...
2026-04-28T08:14:12.821Z [DEBUG] Found 0 plugins (0 enabled, 0 disabled)
2026-04-28T08:14:12.822Z [DEBUG] Total LSP servers loaded: 0
2026-04-28T08:14:12.822Z [DEBUG] [LSP SERVER MANAGER] getAllLspServers returned 0 server(s)
2026-04-28T08:14:12.822Z [DEBUG] LSP manager initialized with 0 servers
2026-04-28T08:14:12.822Z [DEBUG] [STARTUP] MCP configs loaded in 1ms
2026-04-28T08:14:12.823Z [DEBUG] LSP server manager initialized successfully
2026-04-28T08:14:12.823Z [DEBUG] LSP notification handlers registered successfully for all 0 server(s)
2026-04-28T08:14:12.825Z [DEBUG] [STARTUP] Running setup()...
2026-04-28T08:14:12.829Z [DEBUG] Stats cache is up to date
2026-04-28T08:14:12.831Z [DEBUG] Plugin autoupdate: skipped (auto-updater disabled)
2026-04-28T08:14:12.841Z [DEBUG] Loading skills from: managed=C:\ProgramData\ClaudeCode\.claude\skills, user=C:\Users\x\.claude\skills, project=[]
2026-04-28T08:14:12.868Z [DEBUG] installed_plugins.json doesn't exist, returning empty V2 object
2026-04-28T08:14:12.880Z [DEBUG] Error log sink initialized
2026-04-28T08:14:12.882Z [DEBUG] Writing to temp file: C:\Users\x\.claude\todos\7f154e06-524d-4880-95cc-8d1596655284-agent-7f154e06-524d-4880-95cc-8d1596655284.json.tmp.20552.1777364052882
2026-04-28T08:14:12.882Z [DEBUG] Temp file written successfully, size: 2 bytes
2026-04-28T08:14:12.883Z [DEBUG] Renaming C:\Users\ged1sf\.claude\todos\7f154e06-524d-4880-95cc-8d1596655284-agent-7f154e06-524d-4880-95cc-8d1596655284.json.tmp.20552.1777364052882 to C:\Users\x\.claude\todos\7f154e06-524d-4880-95cc-8d1596655284-agent-7f154e06-524d-4880-95cc-8d1596655284.json
2026-04-28T08:14:12.883Z [DEBUG] File C:\Users\x\.claude\todos\7f154e06-524d-4880-95cc-8d1596655284-agent-7f154e06-524d-4880-95cc-8d1596655284.json written atomically
2026-04-28T08:14:12.889Z [DEBUG] getPluginSkills: Processing 0 enabled plugins
2026-04-28T08:14:12.890Z [DEBUG] Total plugin skills loaded: 0
2026-04-28T08:14:12.890Z [DEBUG] Total plugin commands loaded: 0
2026-04-28T08:14:12.891Z [DEBUG] Registered 0 hooks from 0 plugins
2026-04-28T08:14:12.908Z [DEBUG] [STARTUP] setup() completed in 83ms
2026-04-28T08:14:12.909Z [DEBUG] [STARTUP] Loading commands and agents...
2026-04-28T08:14:12.911Z [DEBUG] Total plugin agents loaded: 0
2026-04-28T08:14:12.911Z [DEBUG] Loaded 0 unique skills (managed: 0, user: 0, project: 0, legacy commands: 0)
2026-04-28T08:14:12.912Z [DEBUG] getSkills returning: 0 skill dir commands, 0 plugin skills, 0 bundled skills
2026-04-28T08:14:12.913Z [DEBUG] [STARTUP] Commands and agents loaded in 4ms
2026-04-28T08:14:12.913Z [DEBUG] [STARTUP] Running showSetupScreens()...
2026-04-28T08:14:12.915Z [DEBUG] [render] initYoga starting
2026-04-28T08:14:12.943Z [DEBUG] [Claude in Chrome] Found chrome profiles: Default
2026-04-28T08:14:12.967Z [DEBUG] [render] initYoga complete
2026-04-28T08:14:12.976Z [DEBUG] [keybindings] KeybindingSetup initialized with 45 bindings, 0 warnings
2026-04-28T08:14:12.997Z [DEBUG] [Perfetto] initializePerfettoTracing called, env value: undefined
2026-04-28T08:14:13.010Z [DEBUG] [STARTUP] showSetupScreens() completed in 96ms
2026-04-28T08:14:13.023Z [DEBUG] Initialized versioned plugins system with 0 plugins
2026-04-28T08:14:13.030Z [DEBUG] Writing to temp file: C:\Users\ged1sf\.claude.json.tmp.20552.1777364053030
2026-04-28T08:14:13.030Z [DEBUG] Preserving file permissions: 100666
2026-04-28T08:14:13.031Z [DEBUG] Temp file written successfully, size: 4535 bytes
2026-04-28T08:14:13.032Z [DEBUG] Applied original permissions to temp file
2026-04-28T08:14:13.032Z [DEBUG] Renaming C:\Users\x\.claude.json.tmp.20552.1777364053030 to C:\Users\ged1sf\.claude.json
2026-04-28T08:14:13.033Z [DEBUG] File C:\Users\x\.claude.json written atomically
2026-04-28T08:14:13.049Z [DEBUG] Getting matching hook commands for SessionStart with query: startup
2026-04-28T08:14:13.049Z [DEBUG] Found 0 hook matchers in settings
2026-04-28T08:14:13.049Z [DEBUG] Matched 0 unique hooks for query "startup" (0 before deduplication)
2026-04-28T08:14:13.050Z [DEBUG] [render] initYoga starting
2026-04-28T08:14:13.050Z [DEBUG] [render] initYoga complete
2026-04-28T08:14:13.052Z [DEBUG] Ripgrep first use test: PASSED (mode=builtin, path=C:\Develop\node-v24.13.0-win-x64\node_modules\@anthropic-ai\claude-code\vendor\ripgrep\x64-win32\rg.exe)
2026-04-28T08:14:13.056Z [DEBUG] [keybindings] Skipping file watcher - user customization disabled
2026-04-28T08:14:13.059Z [DEBUG] [Claude in Chrome] Found edge profiles: Default
2026-04-28T08:14:13.069Z [DEBUG] [keybindings] KeybindingSetup initialized with 45 bindings, 0 warnings
2026-04-28T08:14:13.133Z [DEBUG] [keybindings] Skipping file watcher - user customization disabled
2026-04-28T08:14:13.133Z [DEBUG] [REPL:mount] REPL mounted, disabled=false
2026-04-28T08:14:13.141Z [DEBUG] Official marketplace auto-install skipped: already_attempted
2026-04-28T08:14:13.142Z [DEBUG] performStartupChecks called
2026-04-28T08:14:13.143Z [DEBUG] Starting background plugin installations
2026-04-28T08:14:13.144Z [DEBUG] performBackgroundPluginInstallations called
2026-04-28T08:14:13.158Z [DEBUG] Setting installation status: 0 marketplaces, 0 installable plugins, 0 uninstallable plugins
2026-04-28T08:14:13.159Z [DEBUG] Loaded plugins - Enabled: 0, Disabled: 0, Commands: 0, Agents: 0, Errors: 0
2026-04-28T08:14:13.161Z [DEBUG] Git remote URL: null
2026-04-28T08:14:13.161Z [DEBUG] No git remote URL found
2026-04-28T08:14:13.162Z [DEBUG] Not in a GitHub repository, skipping path mapping update
2026-04-28T08:14:13.170Z [DEBUG] [Claude in Chrome] Extension not found in any browser
2026-04-28T08:14:13.172Z [DEBUG] Writing to temp file: C:\Users\ged1sf\.claude.json.tmp.20552.1777364053172
2026-04-28T08:14:13.173Z [DEBUG] Preserving file permissions: 100666
2026-04-28T08:14:13.173Z [DEBUG] Temp file written successfully, size: 4535 bytes
2026-04-28T08:14:13.174Z [DEBUG] Applied original permissions to temp file
2026-04-28T08:14:13.174Z [DEBUG] Renaming C:\Users\x\.claude.json.tmp.20552.1777364053172 to C:\Users\ged1sf\.claude.json
2026-04-28T08:14:13.175Z [DEBUG] File C:\Users\x\.claude.json written atomically
2026-04-28T08:14:13.178Z [DEBUG] Session index: added 1, updated 0, removed 0 (total: 23)
2026-04-28T08:14:13.518Z [DEBUG] AutoUpdaterWrapper: Installation type: unknown
2026-04-28T08:14:15.873Z [DEBUG] LSP Diagnostics: getLSPDiagnosticAttachments called
2026-04-28T08:14:15.874Z [DEBUG] LSP Diagnostics: Checking registry - 0 pending
2026-04-28T08:14:15.875Z [DEBUG] Hooks: Found 0 total hooks in registry
2026-04-28T08:14:15.875Z [DEBUG] Hooks: checkForNewResponses returning 0 responses
2026-04-28T08:14:16.037Z [DEBUG] Getting matching hook commands for UserPromptSubmit with query: undefined
2026-04-28T08:14:16.037Z [DEBUG] Found 0 hook matchers in settings
2026-04-28T08:14:16.037Z [DEBUG] Matched 0 unique hooks for query "no match query" (0 before deduplication)
2026-04-28T08:14:16.045Z [DEBUG] FileHistory: Added snapshot for fa85c686-4c8b-45cd-8663-07292f4af9df, tracking 0 files
2026-04-28T08:14:16.100Z [DEBUG] Tool search disabled: MCPSearchTool is not available (may have been disallowed via disallowedTools).
2026-04-28T08:14:16.103Z [DEBUG] [API:request] Creating client, ANTHROPIC_CUSTOM_HEADERS present: false, has Authorization header: false
2026-04-28T08:14:16.104Z [DEBUG] [API:auth] OAuth token check starting
2026-04-28T08:14:16.105Z [DEBUG] [API:auth] OAuth token check complete
2026-04-28T08:14:16.106Z [DEBUG] Total plugin output styles loaded: 0
2026-04-28T08:14:16.132Z [DEBUG] Auto tool search disabled: 0 chars (threshold: 50000, 10% of context)
2026-04-28T08:14:16.137Z [DEBUG] Skills and commands included in Skill tool:
2026-04-28T08:14:16.147Z [DEBUG] [API:request] Creating client, ANTHROPIC_CUSTOM_HEADERS present: false, has Authorization header: false
2026-04-28T08:14:16.148Z [DEBUG] [API:auth] OAuth token check starting
2026-04-28T08:14:16.148Z [DEBUG] [API:auth] OAuth token check complete
2026-04-28T08:14:18.608Z [ERROR] Error in non-streaming fallback: 404 Resource not found
2026-04-28T08:14:19.371Z [ERROR] Error in non-streaming fallback: 404 Resource not found
2026-04-28T08:14:19.397Z [DEBUG] Getting matching hook commands for Stop with query: undefined
2026-04-28T08:14:19.398Z [DEBUG] Found 0 hook matchers in settings
2026-04-28T08:14:19.398Z [DEBUG] Matched 0 unique hooks for query "no match query" (0 before deduplication)
2026-04-28T08:14:19.491Z [DEBUG] Writing to temp file: C:\Users\ged1sf\.claude.json.tmp.20552.1777364059491
2026-04-28T08:14:19.492Z [DEBUG] Preserving file permissions: 100666
2026-04-28T08:14:19.493Z [DEBUG] Temp file written successfully, size: 4558 bytes
2026-04-28T08:14:19.493Z [DEBUG] Applied original permissions to temp file
2026-04-28T08:14:19.493Z [DEBUG] Renaming C:\Users\x\.claude.json.tmp.20552.1777364059491 to C:\Users\ged1sf\.claude.json
2026-04-28T08:14:19.494Z [DEBUG] File C:\Users\x\.claude.json written atomically
2026-04-28T08:14:19.513Z [DEBUG] Writing to temp file: C:\Users\ged1sf\.claude.json.tmp.20552.1777364059513
2026-04-28T08:14:19.513Z [DEBUG] Preserving file permissions: 100666
2026-04-28T08:14:19.514Z [DEBUG] Temp file written successfully, size: 4589 bytes
2026-04-28T08:14:19.514Z [DEBUG] Applied original permissions to temp file
2026-04-28T08:14:19.515Z [DEBUG] Renaming C:\Users\x\.claude.json.tmp.20552.1777364059513 to C:\Users\ged1sf\.claude.json
2026-04-28T08:14:19.515Z [DEBUG] File C:\Users\x\.claude.json written atomically
2026-04-28T08:15:19.466Z [DEBUG] Getting matching hook commands for Notification with query: idle_prompt
2026-04-28T08:15:19.467Z [DEBUG] Found 0 hook matchers in settings
2026-04-28T08:15:19.467Z [DEBUG] Matched 0 unique hooks for query "idle_prompt" (0 before deduplication)". What is the reason that I'm not able to connect to my local model?`

### Expected behavior

It should go through successfully and not crash the container

### Affected layer

unsure

### Additional context

_No response_

--- a/src/semantic-router/pkg/extproc/processor_res_memory.go
+++ b/src/semantic-router/pkg/extproc/processor_res_memory.go
@@ -2,7 +2,6 @@ package extproc
 
 import (
 	"context"
-	"runtime/debug"
 
 	"github.com/vllm-project/semantic-router/src/semantic-router/pkg/observability/logging"
 )
@@ -25,12 +24,11 @@ func (r *OpenAIRouter) scheduleResponseMemoryStore(ctx *RequestContext, response
 
 	currentUserMessage := extractCurrentUserMessage(ctx)
 	currentAssistantResponse := extractAssistantResponseText(responseBody)
-	go func() {
-		defer func() {
-			if rec := recover(); rec != nil {
-				logging.Errorf("Memory store goroutine: recovered panic: %v\n%s", rec, debug.Stack())
-			}
-		}()
+	// goSafely wraps the goroutine in a deferred recover so a panic in
+	// the memory-store path (e.g. an unexpected payload shape) is
+	// logged via observability rather than aborting the router
+	// process (#1843).
+	goSafely("memory_store", func() {
 		bgCtx := context.Background()
 		sessionID, userID, history, err := extractMemoryInfo(ctx)
 		if err != nil {
@@ -57,5 +55,5 @@ func (r *OpenAIRouter) scheduleResponseMemoryStore(ctx *RequestContext, response
 		); err != nil {
 			logging.Warnf("Memory store failed: %v", err)
 		}
-	}()
+	})
 }

--- a/src/semantic-router/pkg/extproc/req_filter_modality.go
+++ b/src/semantic-router/pkg/extproc/req_filter_modality.go
@@ -267,8 +267,12 @@ func (r *OpenAIRouter) executeBoth(ctx *RequestContext, cfg *config.RouterConfig
 	)
 
 	// --- AR text call (parallel) ---
+	// goSafely catches panics in the upstream model client so a
+	// crash there does not take the whole router process down.
+	// wg.Done() is registered before the model call, so it still
+	// fires on panic via the inner defer (#1843).
 	wg.Add(1)
-	go func() {
+	goSafely("modality_ar_text", func() {
 		defer wg.Done()
 		defer func() {
 			if rec := recover(); rec != nil {
@@ -277,11 +281,11 @@ func (r *OpenAIRouter) executeBoth(ctx *RequestContext, cfg *config.RouterConfig
 			}
 		}()
 		textResp, textErr = r.callARModel(ctx, cfg, openAIRequest, arModel)
-	}()
+	})
 
 	// --- Diffusion image call (parallel) ---
 	wg.Add(1)
-	go func() {
+	goSafely("modality_diffusion", func() {
 		defer wg.Done()
 		defer func() {
 			if rec := recover(); rec != nil {
@@ -290,7 +294,7 @@ func (r *OpenAIRouter) executeBoth(ctx *RequestContext, cfg *config.RouterConfig
 			}
 		}()
 		imgRes, imgErr = r.generateImage(ctx, cfg, diffusionModel)
-	}()
+	})
 
 	wg.Wait()
 

--- a/src/semantic-router/pkg/extproc/router_selection.go
+++ b/src/semantic-router/pkg/extproc/router_selection.go
@@ -2,7 +2,6 @@ package extproc
 
 import (
 	"context"
-	"runtime/debug"
 	"time"
 
 	candle_binding "github.com/vllm-project/semantic-router/candle-binding"
@@ -373,12 +372,10 @@ func populateFromReplay(storage lookuptable.LookupTableStorage, reader store.Rea
 // The returned cancel function stops the goroutine.
 func startLookupTablePopulator(storage lookuptable.LookupTableStorage, reader store.Reader, interval time.Duration) func() {
 	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		defer func() {
-			if rec := recover(); rec != nil {
-				logging.Errorf("Lookup table populator goroutine: recovered panic: %v\n%s", rec, debug.Stack())
-			}
-		}()
+	// goSafely so a panic in populateFromReplay (e.g. malformed
+	// replay-store entry) is logged instead of crashing the whole
+	// router process — see #1843.
+	goSafely("lookup_table_populator", func() {
 		ticker := time.NewTicker(interval)
 		defer ticker.Stop()
 		for {
@@ -389,7 +386,7 @@ func startLookupTablePopulator(storage lookuptable.LookupTableStorage, reader st
 				return
 			}
 		}
-	}()
+	})
 	logging.ComponentEvent("extproc", "lookuptable_populator_started", map[string]interface{}{
 		"interval": interval.String(),
 	})

--- a/src/semantic-router/pkg/extproc/safego.go
+++ b/src/semantic-router/pkg/extproc/safego.go
@@ -1,0 +1,41 @@
+// Package extproc — small helper for goroutine panic recovery.
+//
+// Several extproc background goroutines (config-reload debouncer, async
+// memory store, parallel AR text + diffusion calls, lookup-table
+// populator) historically had no panic recovery. A panic inside any of
+// them propagated up and crashed the whole router process with no log
+// line — the symptom reported in
+// https://github.com/vllm-project/semantic-router/issues/1843, where a
+// ~4000-token prompt silently took the container down.
+//
+// `goSafely(name, fn)` wraps `go fn()` with a deferred recover that
+// logs the panic + a short stack snippet via the existing structured
+// logger and returns. The caller's normal error-handling path is
+// unchanged for non-panic outcomes.
+
+package extproc
+
+import (
+	"runtime/debug"
+
+	"github.com/vllm-project/semantic-router/src/semantic-router/pkg/observability/logging"
+)
+
+// goSafely launches fn in a new goroutine with a deferred panic
+// recovery that emits a structured `extproc:goroutine_panic` event
+// instead of letting the panic abort the process. `name` identifies
+// the call site in the resulting log entry.
+func goSafely(name string, fn func()) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logging.ComponentErrorEvent("extproc", "goroutine_panic", map[string]interface{}{
+					"goroutine": name,
+					"panic":     r,
+					"stack":     string(debug.Stack()),
+				})
+			}
+		}()
+		fn()
+	}()
+}

--- a/src/semantic-router/pkg/extproc/safego_test.go
+++ b/src/semantic-router/pkg/extproc/safego_test.go
@@ -1,0 +1,82 @@
+package extproc
+
+import (
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+// Regression for https://github.com/vllm-project/semantic-router/issues/1843.
+//
+// Before this change the four background goroutines listed in the
+// issue had no panic recovery and a panic in any of them aborted the
+// router process with no log line. goSafely is the helper added to
+// extproc/safego.go; the tests below assert that it
+//
+//   1. runs the user function on a separate goroutine,
+//   2. swallows any panic the user function emits, and
+//   3. does not leak that panic to the parent goroutine.
+
+func TestGoSafelyRunsTheFunction(t *testing.T) {
+	var ran atomic.Int32
+	done := make(chan struct{})
+	goSafely("test_runs_fn", func() {
+		ran.Add(1)
+		close(done)
+	})
+
+	select {
+	case <-done:
+	case <-time.After(2 * time.Second):
+		t.Fatal("goSafely did not run the user function within 2s")
+	}
+
+	if got := ran.Load(); got != 1 {
+		t.Fatalf("goSafely should have run fn exactly once, got %d", got)
+	}
+}
+
+func TestGoSafelyRecoversFromPanic(t *testing.T) {
+	// If goSafely lets the panic propagate the test process aborts
+	// with an unrecovered panic. The test PASSING is the assertion.
+	finished := make(chan struct{})
+	goSafely("test_panics", func() {
+		defer close(finished)
+		panic("simulated upstream model crash")
+	})
+
+	select {
+	case <-finished:
+	case <-time.After(2 * time.Second):
+		t.Fatal("panicking goroutine never reached its defer")
+	}
+
+	// Give the deferred recover one extra tick to log + return.
+	time.Sleep(20 * time.Millisecond)
+}
+
+func TestGoSafelyDoesNotBlockCaller(t *testing.T) {
+	// The wrapper must launch its own goroutine; it must not run the
+	// function inline in the caller's goroutine, otherwise a slow
+	// model client would block the request handler.
+	caller := make(chan struct{})
+	stillBlocked := make(chan struct{})
+	goSafely("test_async", func() {
+		<-caller
+		close(stillBlocked)
+	})
+
+	select {
+	case <-stillBlocked:
+		t.Fatal("goSafely should not block the caller until fn signals back")
+	case <-time.After(50 * time.Millisecond):
+		// Expected — the wrapped fn is waiting on `caller`.
+	}
+
+	close(caller)
+	select {
+	case <-stillBlocked:
+	case <-time.After(2 * time.Second):
+		t.Fatal("goSafely never let fn complete after the caller signal")
+	}
+}

--- a/src/semantic-router/pkg/extproc/server.go
+++ b/src/semantic-router/pkg/extproc/server.go
@@ -9,7 +9,6 @@ import (
 	"os"
 	"os/signal"
 	"path/filepath"
-	"runtime/debug"
 	"strings"
 	"sync/atomic"
 	"syscall"
@@ -434,16 +433,14 @@ func (s *Server) watchConfigAndReload(ctx context.Context) {
 							"event":    ev.Op.String(),
 							"delay_ms": 300,
 						})
-						// Slight delay to let file settle
-						go func() {
-							defer func() {
-								if rec := recover(); rec != nil {
-									logging.Errorf("Config reload goroutine: recovered panic: %v\n%s", rec, debug.Stack())
-								}
-							}()
+						// Slight delay to let file settle. Wrap with
+						// goSafely so a panic inside reload() is logged
+						// via the structured `extproc:goroutine_panic`
+						// event instead of crashing the process (#1843).
+						goSafely("config_reload_debouncer", func() {
 							time.Sleep(300 * time.Millisecond)
 							reload()
-						}()
+						})
 					} else {
 						logging.ComponentDebugEvent("extproc", "config_reload_debounced", map[string]interface{}{
 							"file": ev.Name,

